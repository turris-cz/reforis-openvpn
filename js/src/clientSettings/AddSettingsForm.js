/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import {
    useAPIPost,
    Button,
    useAlert,
    FileInput,
    useForm,
    API_STATE,
    formFieldsSize,
} from "foris";

import API_URLs from "API";

export default function AddSettingsForm() {
    const [setAlert] = useAlert();

    const [postSettingsResponse, postSettings] = useAPIPost(
        API_URLs.clientSettings,
        "multipart/form-data"
    );
    useEffect(() => {
        if (postSettingsResponse.state === API_STATE.ERROR) {
            setAlert(postSettingsResponse.data);
        }
    }, [postSettingsResponse, setAlert]);

    const [formState, formChangeHandler, reloadForm] = useForm(validator);
    const formData = formState.data;
    const formErrors = formState.errors || {};
    useEffect(() => {
        reloadForm({ settings: undefined });
    }, [reloadForm]);

    function handleSubmit(event) {
        event.preventDefault();
        const postData = new FormData();
        postData.append("settings", formData.settings);
        postSettings({ data: postData });
    }

    if (!formData) {
        return null;
    }

    return (
        <>
            <h3>{_("Add settings")}</h3>
            <p>
                {_(
                    "Please select a file with OpenVPN client settings you wish to add. New settings will be enabled after uploading. Please note that network will be restarted automatically."
                )}
            </p>
            <form onSubmit={handleSubmit} className={formFieldsSize}>
                <FileInput
                    label={
                        formData.settings
                            ? formData.settings.name
                            : _("Choose settings file...")
                    }
                    files={[formData.settings]}
                    error={formErrors.settings}
                    onChange={formChangeHandler((value) => ({
                        settings: { $set: value },
                    }))}
                    accept=".conf,.ovpn"
                />
                <div className="text-right">
                    <Button
                        type="submit"
                        forisFormSize
                        disabled={formErrors.settings || !formData.settings}
                    >
                        {_("Upload settings")}
                    </Button>
                </div>
            </form>
        </>
    );
}

function validator(formData) {
    // Ignore empty file - submit will be disabled anyways
    if (!formData.settings) {
        return undefined;
    }

    if (formData.settings.size > 1024 * 1024) {
        return { settings: _("File is too big. Maximum size is 1 MB") };
    }

    const filename = formData.settings.name;
    if (filename.length === 0 || filename.length > 50) {
        return {
            settings: _(
                "Filename must be at least 1 and at most 50 characters long."
            ),
        };
    }
    if (!/^[a-zA-Z0-9_.-]*$/.test(filename)) {
        return {
            settings: _(
                "Filename can contain only alphanumeric characters, dots, dashes and underscores."
            ),
        };
    }

    return undefined;
}
