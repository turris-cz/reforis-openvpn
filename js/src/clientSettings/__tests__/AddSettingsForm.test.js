/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render,
    fireEvent,
    wait,
    getByText,
    getByLabelText,
} from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from "jest-mock-axios";

import AddSettingsForm from "../AddSettingsForm";

function getFormElements(componentContainer) {
    return {
        fileInput: getByLabelText(
            componentContainer,
            "Choose settings file..."
        ),
        submitButton: getByText(componentContainer, "Upload settings"),
    };
}

function setFilesProperty(input, files) {
    // Files can't be set by fireEvent, see https://github.com/testing-library/react-testing-library/issues/93
    Object.defineProperty(input, "files", { value: files, configurable: true });
}

describe("<AddSettingsForm />", () => {
    let componentContainer;

    beforeEach(() => {
        const { container } = render(<AddSettingsForm />);
        componentContainer = container;
    });

    it("should be rendered", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should validate form", () => {
        const { fileInput, submitButton } = getFormElements(componentContainer);

        // Wrong length of file name
        setFilesProperty(fileInput, [new File([], "q".repeat(51))]);
        fireEvent.change(fileInput);
        expect(
            getByText(
                componentContainer,
                "Filename must be at least 1 and at most 50 characters long."
            )
        ).toBeDefined();
        expect(submitButton.disabled).toBe(true);

        // Invalid characters
        setFilesProperty(fileInput, [new File([], "!@#!$@%.conf")]);
        fireEvent.change(fileInput);
        expect(
            getByText(
                componentContainer,
                "Filename can contain only alphanumeric characters, dots, dashes and underscores."
            )
        ).toBeDefined();
        expect(submitButton.disabled).toBe(true);

        // File too big
        setFilesProperty(fileInput, [
            new File(["q".repeat(1024 * 1025)], "turris.conf"),
        ]);
        fireEvent.change(fileInput);
        expect(
            getByText(
                componentContainer,
                "File is too big. Maximum size is 1 MB"
            )
        ).toBeDefined();
        expect(submitButton.disabled).toBe(true);
    });

    it("should add setting", () => {
        const { fileInput, submitButton } = getFormElements(componentContainer);

        expect(submitButton.disabled).toBe(true);
        setFilesProperty(fileInput, [new File([], "turris.conf")]);
        fireEvent.change(fileInput);
        expect(submitButton.disabled).toBe(false);

        fireEvent.click(submitButton);
        expect(mockAxios.post).toBeCalledWith(
            "/reforis/openvpn/api/client-settings",
            expect.anything(), // Can't compare two FormData objects
            expect.anything()
        );
    });

    it("should handle API error", async () => {
        const { fileInput, submitButton } = getFormElements(componentContainer);

        setFilesProperty(fileInput, [new File([], "turris.conf")]);
        fireEvent.change(fileInput);
        fireEvent.click(submitButton);

        const errorMessage = "API didn't handle this well";
        mockJSONError(errorMessage);
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
        });
    });
});
