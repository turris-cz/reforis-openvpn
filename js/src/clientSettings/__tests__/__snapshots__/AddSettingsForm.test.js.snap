// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddSettingsForm /> should be rendered 1`] = `
<div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Add Settings
    </h2>
    <p>
      Please select a file with OpenVPN client settings you wish to add. The new settings will be enabled after the file is uploaded. Note that the network will be restarted automatically.
    </p>
    <form
      class="col px-0"
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label custom-file-label"
          for="1"
        >
          Choose settings file...
        </label>
        <div
          class="input-group custom-file"
        >
          <input
            accept=".conf,.ovpn"
            class="form-control custom-file-input"
            files=""
            id="1"
            type="file"
          />
        </div>
      </div>
      <div
        class="col-sm-12 col-lg-12 p-0 mb-3 text-right"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          disabled=""
          type="submit"
        >
          <span>
            Add
          </span>
        </button>
      </div>
    </form>
  </div>
</div>
`;
